plugins/widget/plugin.js

```diff
@@ -3071,35 +3249,86 @@
 	// LEFT, RIGHT, UP, DOWN, DEL, BACKSPACE - unblock default fake sel handlers.
 	var keystrokesNotBlockedByWidget = { 37: 1, 38: 1, 39: 1, 40: 1, 8: 1, 46: 1 };
 
-	// Applies or removes style's classes from widget.
+	// Adds or removes class to `widget.styleDefinition`.
+	// @param {CKEDITOR.plugins.widget} widget instance.
+	// @param {String} CSS class.
+	// @param {Boolean} whenever to remove or add class.
+	function addRemoveClassToStyleDef( widget, className, remove ) {
+		var classes;
+		if ( widget.styleDefinition && widget.styleDefinition.attributes[ 'class' ] ) {
+			classes = widget.styleDefinition.attributes[ 'class' ].split( /\s+/ );
+		} else {
+			classes = [];
+		}
+
+		if ( remove ) {
+			classes = CKEDITOR.tools.array.filter( classes, function( item ) {
+				return item !== className;
+			} );
+		} else {
+			if ( CKEDITOR.tools.indexOf( classes, className ) == -1 ) {
+				classes.push( className );
+			}
+		}
+		widget.styleDefinition = widget.styleDefinition || { attributes: { 'class': '' } };
+		widget.styleDefinition.attributes[ 'class' ] = classes.join( ' ' );
+	}
+
+	// Applies or removes style's attributes from widget.
 	// @param {CKEDITOR.style} style Custom widget style.
-	// @param {Boolean} apply Whether to apply or remove style.
+	// @param {Boolean} apply Whenever to apply or remove style.
 	function applyRemoveStyle( widget, style, apply ) {
 		var changed = 0,
 			classes = getStyleClasses( style ),
 			updatedClasses = widget.data.classes || {},
-			cl;
-
+			cl,
+			styles = style.getDefinition().styles,
+			attributes = style.getDefinition().attributes;
 		// Ee... Something is wrong with this style.
-		if ( !classes )
+		if ( !classes && !styles && !attributes ) {
 			return;
+		}
 
-		// Clone, because we need to break reference.
-		updatedClasses = CKEDITOR.tools.clone( updatedClasses );
+		if ( classes ) {
+			// Clone, because we need to break reference.
+			updatedClasses = CKEDITOR.tools.clone( updatedClasses );
 
-		while ( ( cl = classes.pop() ) ) {
-			if ( apply ) {
-				if ( !updatedClasses[ cl ] )
-					changed = updatedClasses[ cl ] = 1;
-			} else {
-				if ( updatedClasses[ cl ] ) {
+			while ( ( cl = classes.pop() ) ) {
+				if ( apply ) {
+					if ( !updatedClasses[ cl ] )
+						changed = updatedClasses[ cl ] = 1;
+				} else if ( updatedClasses[ cl ] ) {
 					delete updatedClasses[ cl ];
 					changed = 1;
 				}
 			}
+			if ( changed ) {
+				widget.setData( 'classes', updatedClasses );
+			}
+		}
+
+		applyRemoveStylesOrAttributes( widget, styles, 0, apply );
+		applyRemoveStylesOrAttributes( widget, attributes, 1, apply );
+	}
+
+	// Applies or removes widgets styles/attributes.
+	// @param {CKEDITOR.plugins.widget} Widget instance.
+	// @param {Object} Pairs of inline styles or attributes.
+	// @param {Boolean} whenever to apply or remove style.
```

Comment:
There are four parameters, yet comment describes only three and additionally not in order.